x/proof/keeper/msg_server_submit_proof_test.go

```diff
@@ -0,0 +1,1336 @@
+				return newTestProofMsg(t, supplierAddr, validSessionHeader, wrongPathSessionTree, wrongClosestProofPath)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidProof.Wrapf(
+					"proof path %x does not match block hash %x",
+					wrongClosestProofPath,
+					expectedClosestMerkleProofPath,
+				).Error(),
+			),
+		},
+		{
+			desc: "relay difficulty must be greater than or equal to minimum",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Set the minimum relay difficulty to a non-zero value such that the relays
+				// constructed by the test helpers have a negligable chance of being valid.
+				err := keepers.Keeper.SetParams(ctx, types.Params{
+					MinRelayDifficultyBits: 10,
+				})
+				require.NoError(t, err)
+
+				// Reset the minimum relay difficulty to zero after this test case.
+				t.Cleanup(func() {
+					err := keepers.Keeper.SetParams(ctx, types.DefaultParams())
+					require.NoError(t, err)
+				})
+
+				// Construct a proof message with a session tree containing
+				// a relay of insufficient difficulty.
+				return newTestProofMsg(t,
+					supplierAddr,
+					validSessionHeader,
+					validSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofInvalidRelay.Wrapf(
+					"relay difficulty %d is less than the minimum difficulty %d",
+					validClosestRelayDifficultyBits,
+					10,
+				).Error(),
+			),
+		},
+		{ // group: claim must exist for proof message
+			desc: "claim must exist for proof message",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Construct a new session tree corresponding to the unclaimed session.
+				unclaimedSessionTree := newFilledSessionTree(
+					ctx, t,
+					"wrong_supplier",
+					wrongSupplierAddr,
+					unclaimedSessionHeader,
+					unclaimedSessionHeader,
+					unclaimedSessionHeader,
+					keyRing,
+					ringClient,
+					5,
+				)
+
+				// Discard session tree Merkle root because no claim is being created.
+				// Session tree must be closed (flushed) to compute closest Merkle Proof.
+				_, err = unclaimedSessionTree.Flush()
+				require.NoError(t, err)
+
+				// Construct new proof message using the supplier & session header
+				// from the session which is *not* expected to be claimed.
+				return newTestProofMsg(t,
+					wrongSupplierAddr,
+					unclaimedSessionHeader,
+					unclaimedSessionTree,
+					expectedClosestMerkleProofPath,
+				)
+			},
+			expectedErr: status.Error(
+				codes.FailedPrecondition,
+				types.ErrProofClaimNotFound.Wrapf(
+					"no claim found for session ID %q and supplier %q",
+					unclaimedSessionHeader.GetSessionId(),
+					wrongSupplierAddr,
+				).Error(),
+			),
+		},
+		{
+			desc: "claim and proof session start heights must match",
+			newProofMsg: func(t *testing.T) *types.MsgSubmitProof {
+				// Advance the block height such that no hydration errors can occur when
```

Comment:
I think we need to #PUC. I don't fully understand what this comment is explaining/

WHy would session start height < block height cause an issue?